### PR TITLE
worker/: fix worker stuck in unitTransWaitCondition

### PR DESCRIPTION
**Pull request by lichunzhu**
### What problem does this PR solve? 
https://github.com/pingcap/dm/issues/576
When subtask is started, it will use `unitTransWaitCondition` to wait for relay log to catch up. But if this task is paused or stopped, we should quit as fast as we can instead of keeping waiting.

### What is changed and how it works?
1. Use `st.currCtx` to control wait of `unitTransWaitCondition`.
2. Use parent context `st.ctx` and son context `st.currCtx` to make sure task can be shut down.

### Check List 

Tests 

 - Unit test
 - Integration test

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

File: dm/worker/subtask.go [hunk 1]
```diff
@@ -73,9 +73,13 @@ type SubTask struct {
...
 
 	sync.RWMutex
-	wg     sync.WaitGroup
+	wg sync.WaitGroup
+	// ctx is used for the whole subtask. It will be created only when we new a subtask.
 	ctx    context.Context
 	cancel context.CancelFunc
+	// currCtx is used for one loop. It will be created each time we use st.run/st.Resume
+	currCtx    context.Context
+	currCancel context.CancelFunc
```
Comment (csuzhangxc) [inline dm/worker/subtask.go:82]: Can we do not keep these two fields in the `struct` (to make it clear)?

Comment (lichunzhu) [reply dm/worker/subtask.go:82]: Do you mean use one context?

Comment (csuzhangxc) [reply dm/worker/subtask.go:82]: pass the local variable between methods (it seems not enough?)

Comment (lichunzhu) [reply dm/worker/subtask.go:82]: How to pass local variable between methods?

File: dm/worker/subtask.go [hunk 2]
```diff
@@ -186,36 +193,59 @@ func (st *SubTask) Run() {
...
+	st.setStage(pb.Stage_Running)
+	ctx, cancel := context.WithCancel(st.ctx)
+	st.setCurrCtx(ctx, cancel)
+	err := st.unitTransWaitCondition(ctx)
 	if err != nil {
 		st.l.Error("wait condition", log.ShortError(err))
 		st.fail(err)
 		return
+	} else if ctx.Err() != nil {
+		return
```
Comment (csuzhangxc) [inline dm/worker/subtask.go:205]: Will the stage keep as `Stage_Running`?

File: dm/worker/subtask.go [hunk 3]
```diff
@@ -73,9 +73,13 @@ type SubTask struct {
 	l log.Logger
 
 	sync.RWMutex
-	wg     sync.WaitGroup
+	wg sync.WaitGroup
+	// ctx is used for the whole subtask. It will be created only when we new a subtask.
 	ctx    context.Context
 	cancel context.CancelFunc
+	// currCtx is used for one loop. It will be created each time we use st.run/st.Resume
+	currCtx    context.Context
```
Comment (csuzhangxc) [inline dm/worker/subtask.go:81]: Is it enough to just let `ctx` as `used for one loop`

File: dm/worker/subtask_test.go [hunk 4]
```diff
@@ -479,3 +485,62 @@ func (t *testSubTask) TestSubtaskWithStage(c *C) {
 	c.Assert(st.CurrUnit(), Equals, nil)
 	c.Assert(st.Result(), IsNil)
 }
+
+func (t *testSubTask) TestSubtaskFastQuit(c *C) {
+	// case: test subtask stuck into unitTransWaitCondition
+	cfg := &config.SubTaskConfig{
+		Name: "testSubtaskFastQuit",
+		Mode: config.ModeFull,
```
Comment (csuzhangxc) [inline dm/worker/subtask_test.go:493]: Why not `All`?

File: dm/worker/subtask_test.go [hunk 5]
```diff
@@ -479,3 +485,62 @@ func (t *testSubTask) TestSubtaskWithStage(c *C) {
...
+	mockLoader := NewMockUnit(pb.UnitType_Load)
+	mockSyncer := NewMockUnit(pb.UnitType_Sync)
+
+	st := NewSubTaskWithStage(cfg, pb.Stage_Paused, nil)
+	st.prevUnit = mockLoader
+	st.currUnit = mockSyncer
+
+	finished := make(chan struct{})
+	go func() {
+		close(finished)
```
Comment (csuzhangxc) [inline dm/worker/subtask_test.go:515]: Close directly or do you mean `defer`?

File: dm/worker/subtask.go (hunk 3 above)
Comment (lichunzhu) [reply dm/worker/subtask.go:81]: I don't think it's enough. Suggest a situation:
1. we close a subtask and remove subtask from subtask holder.
2. After this subtask's cancel was runned, a goroutine build a new ctx and cancel and start `subtask.Run`.
If this worker is not stopped, this subtask will keep running without monitor. It's dangerous.

File: dm/worker/subtask.go (hunk 2 above)
Comment (lichunzhu) [reply dm/worker/subtask.go:205]: `ctx.Err() != nil` means this context is canceled in other go routine, that go routine will change the stage.

Comment (csuzhangxc) [reply dm/worker/subtask.go:205]: So, how about checking the stage in the unit test cases?

**Comment by csuzhangxc** [comment]
@lichunzhu please do not forget to cherry-pick it manually.

**Comment by lichunzhu** [comment]
> @lichunzhu please do not forget to cherry-pick it manually.

@csuzhangxc OK

**Pull request merged**